misc/multistream-select/src/dialer_select.rs

```diff
@@ -261,6 +400,168 @@ where
     }
 }
 
+struct SimOpenPhase<R> {
+    state: SimOpenState<R>,
+}
+
+enum SimOpenState<R> {
+    SendNonce { io: MessageIO<R> },
+    FlushNonce { io: MessageIO<R>, local_nonce: u64 },
+    ReadNonce { io: MessageIO<R>, local_nonce: u64 },
+    SendRole { io: MessageIO<R>, local_role: SimOpenRole },
+    FlushRole { io: MessageIO<R>, local_role: SimOpenRole },
+    ReadRole { io: MessageIO<R>, local_role: SimOpenRole },
+    Done,
+}
+
+enum SimOpenRole {
+    Initiator,
+    Responder,
+}
+
+impl<R> Future for SimOpenPhase<R>
+where
+    // The Unpin bound here is required because we produce a `Negotiated<R>` as the output.
+    // It also makes the implementation considerably easier to write.
+    R: AsyncRead + AsyncWrite + Unpin,
+{
+    type Output = Result<(MessageIO<R>, SimOpenRole), NegotiationError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+
+        loop {
+            match mem::replace(&mut self.state, SimOpenState::Done) {
+                SimOpenState::SendNonce { mut io } => {
+                    match Pin::new(&mut io).poll_ready(cx)? {
```

Comment:
shouldn’t this always succeed? (I’m not deep inside libp2p — the last relevant call looks to have been `poll_flush`, whose name seems to suggest that the socket should be ready)

Reply:
I think in libp2p-quic we're ignoring the flush. a substream is flushed on close, but there is no support in quinn to manually flush a stream.

- [0] https://github.com/ipfs-rust/libp2p-quic/blob/master/src/muxer.rs#L366

Reply:
I am not quite sure I follow @rkuhn, `poll_flush` does not imply the same as `poll_ready` (see the [`Sink` docs](https://docs.rs/futures/0.3.15/futures/sink/trait.Sink.html#tymethod.poll_flush)), thus one can not call `start_send` right away. Does that make sense?

Reply:
Re-reading the Sink docs I see what you mean, although the docs are still somewhat weird: I was tripped up by the documentation of `start_send` which basically implies that `poll_flush` is stronger than `poll_ready`. And I have a hard time imagining a useful implementation where `poll_ready` wouldn’t be immediately ready directly after a `poll_flush`.